python/shotgun_model/shotgun_model.py

```diff
@@ -1164,141 +1133,135 @@ def __process_thumbnail_for_item(self, item):
+        # set up default thumb
+        self._populate_default_thumbnail(item)
 
-            # construct tree view node object
-            item = ShotgunStandardItem(dv)
+        # run the populate item method (only runs at construction, not on cache restore)
+        if is_leaf:
+            self._populate_item(item, sg_item)
+        else:
+            self._populate_item(item, None)
 
-            # keep tabs of which items we are creating
-            item.setData(True, self.IS_SG_MODEL_ROLE)
+        # run the finalizer (always runs on construction, even via cache)
+        self._finalize_item(item)
 
-            # keep a reference to this object to make GC happy
-            # (pyside may crash otherwise)
-            self.__all_tree_items.append(item)
+        # request thumb
+        if self.__download_thumbs:
+            self.__process_thumbnail_for_item(item)
 
-            # get the full sg data dict that corresponds to this folder item
-            # note that this item may only partially match the sg data
-            # for leaf item, the sg_item completely matches the item
-            # but higher up it will be a subset of the fields only.
-            sg_item = discrete_values[dv]
+        return item
 
-            # store the actual field value we have for this item
-            item.setData({"name": field, "value": sg_item[field] }, self.SG_ASSOCIATED_FIELD_ROLE)
+    def __realize_parent_child_hierarchy_r(self, node):
+        """
+        Inserts all children nodes into their parent, recursively.
 
-            if on_leaf_level:
-                # this is the leaf level
-                # attach the shotgun data so that we can access it later
-                # note - pyqt converts everything automatically to unicode,
-                # but using somewhat strange rules, so properly convert
-                # values to unicode prior to insertion
-                item.setData(sanitize_for_qt_model(sg_item), self.SG_DATA_ROLE)
+        :param node: Dictionary with keys 'item' and 'children' representing a ShotgunModelItem and a list
+            of ShotgunModelItem to be appended as rows of 'item'.
+        """
+        # If this is a leaf, there are no children.
+        if "children" not in node:
+            return
 
-                # and also populate the id association in our lookup dict
-                self.__entity_tree_data[ sg_item.get("id") ] = item
+        children = node["children"]
+        for child_key in sorted(children.keys(), cmp=lambda x,y: cmp(x.lower(), y.lower())):
+            child = children[child_key]
+            node["item"].appendRow(child["item"])
+            self.__realize_parent_child_hierarchy_r(child)
 
-            # now we got the object set up. Now start calling custom methods:
+    def __populate_complete_tree(self, sg_data):
+        """
+        Generate tree model data structure based on Shotgun data.
```

Comment:
I think it would be cool to add a few sentences here to go through the general approach - explain why you refactored it into a two pass approach like this (for performance). Just some notes to help future readers of the code understand why this particular approach was used!